internal/controllers/core/kubernetesapply/reconciler.go

```diff
@@ -76,23 +83,160 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
+				// If the map isn't found, keep going and let shouldDeployOnReconcile
+				// handle it.
+				continue
+			}
+			return ctrl.Result{}, err
+		}
+
+		imageMaps[nn] = &im
+	}
+
+	if !r.shouldDeployOnReconcile(request.NamespacedName, &ka, imageMaps) {
+		// TODO(nick): Like with other reconcilers, there should always
+		// be a reason why we're not deploying, and we should update the
+		// Status field of KubernetesApply with that reson.
+		return ctrl.Result{}, nil
+	}
+
+	// Update the apiserver with the result of this deploy.
+	_, err = r.ForceApply(ctx, nn, ka.Spec, imageMaps)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// Determine if we should deploy the current YAML.
+//
+// Ensures:
+// 1) We have enough info to deploy, and
+// 2) Either we haven't deployed before,
+//    or one of the inputs has changed since the last deploy.
+func (r *Reconciler) shouldDeployOnReconcile(
+	nn types.NamespacedName,
+	ka *v1alpha1.KubernetesApply,
+	imageMaps map[types.NamespacedName]*v1alpha1.ImageMap) bool {
+	if ka.ObjectMeta.Labels[v1alpha1.LabelOwnerKind] == v1alpha1.LabelOwnerKindTiltfile {
+		// Until resource dependencies are expressed in the API,
+		// we can't use reconciliation to deploy KubernetesApply objects.
+		return false
+	}
+
+	for _, imageMapName := range ka.Spec.ImageMaps {
+		_, ok := imageMaps[types.NamespacedName{Name: imageMapName}]
+		if !ok {
+			// We haven't built the images yet to deploy.
+			return false
+		}
+	}
+
+	r.mu.Lock()
+	result, ok := r.results[nn]
+	r.mu.Unlock()
+
+	if !ok {
+		// We've never successfully deployed before, so deploy now.
+		return true
+	}
+
+	if !apicmp.DeepEqual(ka.Spec, result.Spec) {
+		// The YAML to deploy changed.
+		return true
+	}
+
+	imageMapNames := ka.Spec.ImageMaps
+	if len(imageMapNames) != len(result.ImageMapSpecs) ||
+		len(imageMapNames) != len(result.ImageMapStatuses) {
+		return true
+	}
+
+	for i, name := range ka.Spec.ImageMaps {
+		im := imageMaps[types.NamespacedName{Name: name}]
+		if !apicmp.DeepEqual(im.Spec, result.ImageMapSpecs[i]) {
+
+			return true
+		}
+		if !apicmp.DeepEqual(im.Status, result.ImageMapStatuses[i]) {
+			return true
+		}
+	}
+
+	return false
 }
 
 // Inject the images into the YAML and apply it to the cluster, unconditionally.
 //
+// Update the apiserver when finished.
+//
 // We expose this as a public method as a hack! Currently, in Tilt, BuildController
 // handles dependencies between resources. The API server doesn't know about build
 // dependencies yet. So Tiltfile-owned resources are applied manually, rather than
 // going through the normal reconcile system.
 func (r *Reconciler) ForceApply(
+	ctx context.Context,
+	nn types.NamespacedName,
+	spec v1alpha1.KubernetesApplySpec,
+	imageMaps map[types.NamespacedName]*v1alpha1.ImageMap) (v1alpha1.KubernetesApplyStatus, error) {
+	status := r.forceApplyHelper(ctx, spec, imageMaps)
+
+	result := Result{
+		Spec:   spec,
+		Status: status,
```

Comment:
I think this usage is actually provably safe right now (status has no reference types + based on code paths we never actually use both the return + the version stored in results), but it might be good to `*status.DeepCopy()` here anyway to avoid a future surprise (or incur the wrath of the race detector)

Reply:
done